fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,113 @@
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    *   - create a shallow copy of the todos array from state
+    *   - add the todo received from the server to the copied array
+    *   - replace the old todos array in state with the new one
+    *
+    * DO NOT MUTATE the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+    const stringToDo = JSON.stringify(newTodo);
+    const shallow = this.state.todos;
+    fetch('/api/todos', { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: stringToDo })
+      .then(response => response.json())
+      .then(data => this.setState(
+        {
+          todos: shallow.concat(data)
+        })
+      );
+  }
+
+  toggleCompleted(todoId) {
+    /**
+     * Find the index of the todo with the matching todoId in the state array.
+     * Get its "isCompleted" status.
+     * Make a new object containing ONE PROPERTY: the opposite "isCompleted" status.
+     * Use fetch to send a PATCH request to `/api/todos/${todoId}`
+     * Then 😉, once the response JSON is received and parsed,
+     *   - create a shallow copy of the todos array from state
+     *   - replace the old todo with the todo received from the server
+     *   - replace the old todos in the state with the new one (you know the index).
+     *
+     * NOTE: "toggle" means to flip back and forth, so clicking a todo
+     * in the list repeatedly should "toggle" its isCompleted status back and forth.
+     *
+     * DO NOT try to calculate the index of the todo by subtracting 1 from the id.
+     *
+     * DO NOT MUTATE the original state array, nor any objects within it.
+     * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+     *
+     * TIP: Be sure to SERIALIZE the updates in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     */
+
+    const modObj = this.state.todos.find(item => item.todoId === todoId);
+    // function finders(compare) {
+    //   return compare === modObj;
+    // }
+    // const modObjInd = this.state.todos.findIndex(finders);
+    modObj.isCompleted = !modObj.isCompleted;
+    const JSONmodObj = JSON.stringify(modObj);
+    fetch(`/api/todos/${todoId}`, { method: 'PATCH', headers: { 'Content-Type': 'application/json' }, body: JSONmodObj })
+      .then(response => response.json())
+      .then(data => {
+        const shallow = this.state.todos;
+        this.setState(
+          {
+            todos: shallow
+          }
+        );
```

Comment:
Again this won't create a shallow copy of the array. Here you want to use `map` to create a new array and swap out the updated todo item:
```JAVASCRIPT
const shallow = this.state.todos.map(original => {
  return original.todoId === data.todoId
    ? data
    : original;
});

this.setState(
  {
    todos: shallow
  }
);
```